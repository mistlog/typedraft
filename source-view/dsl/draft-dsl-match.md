```typescript
export class PatternMatch {}
```

```typescript
<PatternMatch /> +
    function Transcribe(block: Array<ExpressionStatement>): [IfStatement] {
        let tail: IfStatement = null;
        let head: IfStatement = null;
        block.forEach(each => {
            const expression = each.expression as ArrowFunctionExpression;
            <TranscribeExpressionToIf />;
        });

        /**
         * transcribed is the "head"(first) of if statement
         */
        return [head];
    };
```

```typescript
function TranscribeExpressionToIf(expression: ArrowFunctionExpression, head: IfStatement, tail: IfStatement) {
    if (expression.params.length === 0) {
        <HandleDefaultCase />;
    } else {
        <BuildCurrentIf />;
        <MoveToNext />;
    }
}
```

```typescript
function BuildCurrentIf(expression: ArrowFunctionExpression) {
    //
    const [pattern_info] = expression.params as [Identifier];
    const to_match = pattern_info.name;
    const annotation = (pattern_info.typeAnnotation as TypeAnnotation).typeAnnotation;

    //
    let current: IfStatement = null;
    if (isTSLiteralType(annotation)) {
        /**
         * number or string
         */
        type Literal = (StringLiteral | NumberLiteral) & { extra: { raw: string } };
        const pattern = (annotation.literal as Literal).extra.raw;
        current = ToAst(`if(${to_match}===${pattern}){}`) as IfStatement;
    } else if (isTSTypeReference(annotation)) {
        /**
         * enum
         */
        current = ToAst(`if(${to_match}===${ToString(annotation)}){}`) as IfStatement;
    }
    current.consequent = expression.body as BlockStatement;
}
```

```typescript
function MoveToNext(head: IfStatement, current: IfStatement, tail: IfStatement) {
    if (head == null) {
        head = tail = current;
    } else {
        tail = tail.alternate = current;
    }
}
```

```typescript
function HandleDefaultCase(expression: ArrowFunctionExpression, tail: IfStatement) {
    tail.alternate = expression.body as BlockStatement;
}
```