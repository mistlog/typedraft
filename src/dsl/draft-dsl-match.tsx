import { Identifier, TypeAnnotation, isTSLiteralType, IfStatement, BlockStatement, StringLiteral, NumberLiteral, isTSTypeReference, ArrowFunctionExpression, Statement, isArrowFunctionExpression, ExpressionStatement, ifStatement } from "@babel/types";
import { ToAst, ToString } from "../common/utility";

export class PatternMatch
{

}

<PatternMatch /> + function Transcribe(block: Array<ExpressionStatement>): [IfStatement]
{
    let tail: IfStatement = null;
    let head: IfStatement = null;

    block.forEach(each =>
    {
        const expression = each.expression as ArrowFunctionExpression;

        //@ts-ignore
        <TranscribeExpressionToIf />;
    })


    /**
     * transcribed is the "head"(first) of if statement
     */
    return [head];
}

function TranscribeExpressionToIf(expression: ArrowFunctionExpression, head: IfStatement, tail: IfStatement)
{
    if (expression.params.length === 0)
    {
        //@ts-ignore
        <HandleDefaultCase />
    }
    else
    {
        //@ts-ignore
        <BuildCurrentIf />;

        //@ts-ignore
        <MoveToNext />;
    }
}

function BuildCurrentIf(expression: ArrowFunctionExpression)
{
    //
    const [pattern_info] = expression.params as [Identifier];
    const to_match = pattern_info.name;
    const annotation = (pattern_info.typeAnnotation as TypeAnnotation).typeAnnotation;

    //
    let current: IfStatement = null;

    if (isTSLiteralType(annotation))
    {
        /**
         * number or string
         */
        type Literal = (StringLiteral | NumberLiteral) & { extra: { raw: string } };
        const pattern = (annotation.literal as Literal).extra.raw;
        current = ToAst(`if(${to_match}===${pattern}){}`) as IfStatement;
    }
    else if (isTSTypeReference(annotation))
    {
        /**
         * enum
         */
        current = ToAst(`if(${to_match}===${ToString(annotation)}){}`) as IfStatement;
    }

    current.consequent = expression.body as BlockStatement;
}

function MoveToNext(head: IfStatement, current: IfStatement, tail: IfStatement)
{
    if (head == null)
    {
        head = tail = current;
    }
    else
    {
        tail = tail.alternate = current;
    }

}


function HandleDefaultCase(expression: ArrowFunctionExpression, tail: IfStatement)
{
    tail.alternate = expression.body as BlockStatement;
}