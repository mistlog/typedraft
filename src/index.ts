export * from "./core/transcriber";

export * from "./code-object/export-class";
export * from "./code-object/local-context";
export * from "./code-object/inline-context";
export * from "./code-object/method";
export * from "./code-object/module";

export * from "./plug-in/draft-plugin-class";
export * from "./plug-in/draft-plugin-dsl";
export * from "./plug-in/draft-plugin-filter";
export * from "./plug-in/draft-plugin-local-context";
export * from "./plug-in/draft-plugin-refresh";

export * from "./common/utility";

/**
 *
 */
import { Transcriber } from "./core/transcriber";
import { RefreshDraftPlugin } from "./plug-in/draft-plugin-refresh";
import { DSLPlugin } from "./plug-in/draft-plugin-dsl";
import { LocalContextPlugin } from "./plug-in/draft-plugin-local-context";
import { ClassPlugin } from "./plug-in/draft-plugin-class";
import { FilterPlugin } from "./plug-in/draft-plugin-filter";

export function MakeDefaultTranscriber(_module: string) {
    const transcriber = new Transcriber(_module);
    transcriber.m_Plugins = [
        new RefreshDraftPlugin(transcriber),
        new DSLPlugin(transcriber),
        new RefreshDraftPlugin(transcriber),
        new LocalContextPlugin(transcriber),
        new ClassPlugin(transcriber),
        new FilterPlugin(transcriber),
    ];
    return transcriber;
}
