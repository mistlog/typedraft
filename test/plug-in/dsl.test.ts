import { MakeDefaultTranscriber, IDSL, ToAst } from "../../src";
import {
    Statement,
    FunctionDeclaration,
    labeledStatement,
    blockStatement,
    identifier,
} from "@babel/types";
import { NodePath } from "@babel/traverse";
import { PatternMatch } from "draft-dsl-match";

test("simple", () => {
    //
    class Foo implements IDSL {
        Transcribe(block: Array<Statement>): Array<Statement> {
            const transcribed = ToAst(`
                    console.log("current");
                `);

            return [transcribed as Statement];
        }
    }

    //
    const code = `
        export function Main(){
            <Test/>;
        }

        function Test(){
            'use foo';
            console.log("previous");
        }
    `;
    const transcriber = MakeDefaultTranscriber(code);
    transcriber.AddDSL("foo", new Foo());
    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

/**
 * <TestLog/> is not used as local context, but after DSL "match" resolved,
 * it will be used as local context,
 * thus after DSL resolved, we need to refresh draft again.
 */
test("local context added after dsl resolved", () => {
    const code = `
            export function Main(){
                <TestMatch/>;
            }

            function TestMatch(value){
                "use match";

                (value: "a") => <TestLog/>;
            }

            function TestLog(){
                console.log("value is a");
            }
        `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.AddDSL("match", new PatternMatch());
    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

test("nested dsl", () => {
    //
    class Watch implements IDSL {
        Transcribe(block: Array<Statement>): Array<Statement> {
            return [labeledStatement(identifier("$"), blockStatement(block))];
        }
    }

    //
    const code = `
            export function Main(){
                <Test/>;
            }

            function Test(value: number){
                "use watch";

                <Match/>;
                b = value + 1;
            }

            function Match(value){
                "use match";

                (value: "a") =>
                {
                    console.log("value is a");
                };
            }
        `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.AddDSL("watch", new Watch());
    transcriber.AddDSL("match", new PatternMatch());
    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

test("use path param", () => {
    //
    class Foo implements IDSL {
        Transcribe(block: Array<Statement>, path: NodePath<FunctionDeclaration>): Array<Statement> {
            path.traverse({
                Identifier(path) {
                    if (path.node.name === "x") {
                        path.node.name = "y";
                    }
                },
            });

            return block;
        }
    }

    //
    const code = `
            export function Main(){
                <Test/>;
            }

            function Test(){
                'use foo';
                const x = 1;
                console.log(x + 1);
            }
        `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.AddDSL("foo", new Foo());
    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});
