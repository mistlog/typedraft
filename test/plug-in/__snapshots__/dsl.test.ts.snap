// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inline context 1`] = `
"export function Main() {
  console.log(\\"hello\\");
  {
    console.log(\\"current\\");
  }
  console.log(\\"world\\");
}"
`;

exports[`inline context: merge 1`] = `
"export function Main() {
  console.log(\\"hello\\");
  console.log(\\"current\\");
  console.log(\\"world\\");
}"
`;

exports[`local context added after dsl resolved 1`] = `
"export function Main() {
  if (value === \\"a\\") {
    console.log(\\"value is a\\");
  }
}"
`;

exports[`nested dsl 1`] = `
"export function Main() {
  $: {
    if (value === \\"a\\") {
      console.log(\\"value is a\\");
    }

    b = value + 1;
  }
}"
`;

exports[`nested dsl: merge 1`] = `
"export function Main() {
  $: {
    if (value === \\"a\\") {
      console.log(\\"value is a\\");
    }

    b = value + 1;
  }
}"
`;

exports[`no dsl 1`] = `
"export function Main() {
  console.log(\\"previous\\");
}"
`;

exports[`simple 1`] = `
"export function Main() {
  console.log(\\"current\\");
}"
`;

exports[`use path param 1`] = `
"export function Main() {
  const y = 1;
  console.log(y + 1);
}"
`;
